src/web-auth/captcha.js

```diff
@@ -0,0 +1,117 @@
+    'recaptcha_v2': function () {
+      return '<div class= "recaptcha" ></div><input type="hidden" name="captcha" />';
+    }
+  }
+};
+
+function handleAuth0Provider(element, options, challenge, load) {
+  element.innerHTML = options.templates[challenge.provider](challenge);
+  element.querySelector('.captcha-reload').addEventListener('click', function (e) {
+    e.preventDefault();
+    load();
+  });
+}
+
+function injectRecaptchaScript(element, lang, callback) {
+  var callbackName = 'recatpchaCallback_' + Math.floor(Math.random() * 1000001);
+  window[callbackName] = function () {
+    delete window[callbackName];
+    callback();
+  };
+  var script = window.document.createElement('script');
+  script.src = 'https://www.google.com/recaptcha/api.js?hl=' + lang + '&onload=' + callbackName;
+  script.async = true;
+  window.document.body.appendChild(script);
+}
+
+function handleRecaptchaProvider(element, options, challenge) {
+  var widgetId = element.hasAttribute('data-wid') && element.getAttribute('data-wid');
+
+  function setValue(value) {
+    var input = element.querySelector('input[name="captcha"]');
+    input.value = value || '';
+  }
+
+  if (widgetId) {
+    setValue();
+    window.grecaptcha.reset(widgetId);
+    return;
+  }
+
+  element.innerHTML = options.templates[challenge.provider](challenge);
+
+  var recaptchaDiv = element.querySelector('.recaptcha');
+
+  injectRecaptchaScript(element, options.lang, function () {
+    widgetId = window.grecaptcha.render(recaptchaDiv, {
+      callback: setValue,
+      'expired-callback': function () { setValue(); },
+      'error-callback': function () { setValue(); },
+      sitekey: challenge.siteKey
+    });
+    element.setAttribute('data-wid', widgetId)
+  });
+}
+
+
+/**
+ *
+ * Renders the captcha challenge in the provided element.
+ *
+ * @param {Authentication} auth0Client The challenge response from the authentication server
+ * @param {HTMLElement} element The element where the captcha needs to be rendered
+ * @param {Object} params The configuration params for the captcha
+ * @param {Object} [params.templates] An object containaing templates for each captcha provider
+ * @param {Function} [params.templates.auth0] template function receiving the challenge and returning an string
+ * @param {Function} [params.templates.recaptcha_v2] template function receiving the challenge and returning an string
+ * @param {String} [params.lang=en] the ISO code of the language for recaptcha
+ */
+function render(auth0Client, element, options) {
+  options = object.merge(defaults).with(options || {});
+
+  function load() {
+    auth0Client.getChallenge(function (err, challenge) {
+      if (err) { return; }
+      if (!challenge.required) {
+        element.style.display = 'none';
+        element.innerHTML = '';
+        return;
+      }
+      element.style.display = '';
+      if (challenge.provider === 'auth0') {
+        handleAuth0Provider(element, options, challenge, load);
+      }
```

Comment:
should we return or use else-if here? in case someone misses this in the future and adds logic below this that will get executed

Reply:
thanks!